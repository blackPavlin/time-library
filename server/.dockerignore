dist
node_modules
tests
.eslint
.gitignore
jest.config.js
tsconfig.eslint.json
*.md
