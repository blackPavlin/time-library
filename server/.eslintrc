{
    "root": true,
    "env": {
        "node": true,
        "jest": true,
        "es2020": true
    },
    "extends": [
        "prettier",
        "eslint:recommended",
        "plugin:prettier/recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/eslint-recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking"
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "project": "./tsconfig.json",
        "tsconfigRootDir": "."
    },
    "plugins": [
        "@typescript-eslint",
        "prettier",
        "jest"
    ],
    "ignorePatterns": ["*.js"],
    "rules": {
        "@typescript-eslint/no-floating-promises": "off",
        "@typescript-eslint/unbound-method": "off",
        "prettier/prettier": [
            "error",
            {
                "trailingComma": "all",
                "singleQuote": true,
                "printWidth": 100,
                "useTabs": true
            }
        ]
    }
}
