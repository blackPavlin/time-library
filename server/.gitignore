node_modules
dist
local.config.env
