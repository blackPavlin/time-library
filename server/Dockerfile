FROM node:alpine as build

WORKDIR /build
COPY package*.json ./
RUN npm install
COPY . .
RUN npm run build

FROM node:alpine

WORKDIR /root
COPY --from=build /build/package*.json ./
RUN npm install --only=production
COPY --from=build /build/dist ./build/
COPY --from=build /build/env ./env/
EXPOSE 3000
CMD ["node", "./build/app"]
