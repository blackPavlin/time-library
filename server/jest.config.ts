import { InitialOptionsTsJest } from 'ts-jest';

const config: InitialOptionsTsJest = {
	testEnvironment: 'node',
	testTimeout: 5000,
};

export default config;
