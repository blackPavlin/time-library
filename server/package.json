{
  "name": "server",
  "version": "1.0.0",
  "description": "",
  "main": "dist/app.js",
  "scripts": {
    "start": "node ./dist/app.js",
    "start:dev": "ts-node ./src/app.ts",
    "start:watch": "nodemon ./src/app.ts",
    "build": "tsc",
    "test": "jest --colors",
    "lint": "tsc --noEmit && eslint ./src/**/*.ts",
    "lint:fix": "eslint ./src/**/*.ts --fix"
  },
  "nodemonConfig": {
    "delay": "5000"
  },
  "keywords": [],
  "repository": {
    "type": "git",
    "url": "https://github.com/blackPavlin/fastify-http-server.git"
  },
  "author": {
    "name": "Boikiv Bogdan",
    "email": "boikiv.bogdan@gmail.com"
  },
  "license": "UNLICENSE",
  "dependencies": {
    "bcrypt": "^5.0.1",
    "fastify": "^3.15.1",
    "fastify-auth": "^1.1.0",
    "fastify-cors": "^6.0.1",
    "fastify-env": "^2.1.1",
    "fastify-plugin": "^3.0.0",
    "fastify-sensible": "^3.1.1",
    "json-schema-to-ts": "^1.6.4",
    "jsonwebtoken": "^8.5.1",
    "mongoose": "^5.12.10"
  },
  "devDependencies": {
    "@types/bcrypt": "^5.0.0",
    "@types/jest": "^26.0.23",
    "@types/jsonwebtoken": "^8.5.1",
    "@types/node": "^15.6.0",
    "@typescript-eslint/eslint-plugin": "^4.24.0",
    "@typescript-eslint/parser": "^4.24.0",
    "eslint": "^7.27.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-jest": "^24.3.6",
    "eslint-plugin-prettier": "^3.4.0",
    "jest": "^26.6.3",
    "lint-staged": "^11.0.0",
    "mongodb-memory-server": "^6.9.6",
    "nodemon": "^2.0.7",
    "prettier": "^2.3.0",
    "ts-jest": "^26.5.6",
    "ts-node": "^9.1.1",
    "typescript": "^4.2.4"
  },
  "lint-staged": {
    "*.ts": [
      "npm run lint",
      "npm run test"
    ]
  },
  "husky": {
    "hooks": {
      "pre-commit": "lint-staged"
    }
  }
}
