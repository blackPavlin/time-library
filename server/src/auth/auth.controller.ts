import { Body, Controller, HttpCode, Post } from '@nestjs/common';
import { CreateUserDto } from 'src/user/dto/create-user.dto';
import { AuthService } from './auth.service';

@Controller('auth')
export class AuthController {
  constructor(private authService: AuthService) {}

  @Post('/login')
  @HttpCode(200)
  async login(
    @Body() createUserDto: CreateUserDto,
  ): Promise<{ token: string }> {
    return this.authService.login(createUserDto.login, createUserDto.password);
  }

  @Post('/signup')
  async signup(@Body() createUserDto: CreateUserDto): Promise<void> {
    return this.authService.signup(createUserDto.login, createUserDto.password);
  }
}
