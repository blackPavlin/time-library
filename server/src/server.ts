import path from 'path';
import fastify from 'fastify';
import fastifyCors from 'fastify-cors';
import fastifySensible from 'fastify-sensible';
import fastifyAuth from 'fastify-auth';
import fastifyEnv from 'fastify-env';

import configSchema from './Config';
import authPlugin from './Plugins/auth.plugin';
import authController from './Controllers/auth.controller';
import taskController from './Controllers/task.controller';

const server = fastify({
	logger: {
		level: 'error',
	},
	ignoreTrailingSlash: true,
});

server
	.register(fastifyCors, { origin: '*', methods: ['GET', 'POST', 'PUT', 'DELETE'] })
	.register(fastifySensible)
	.register(fastifyAuth)
	.register(fastifyEnv, {
		schema: configSchema,
		dotenv: { path: path.join(__dirname, '../env/local.config.env') },
	})
	.after(() => {
		server
			.register(authPlugin)
			.register(authController, { prefix: '/api/auth' })
			.register(taskController, { prefix: '/api/task' });
	});

export default server;
