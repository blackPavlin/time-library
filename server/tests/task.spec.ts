import server from '../src/server';
import Tasks from '../src/Models/task.model';
import * as database from './helpers/database';
import * as auth from './helpers/auth';
import * as task from './helpers/task';

const testLogin = 'admin@admin.com';
let taskId = '';

describe('Тест CRUD карточек', () => {
	beforeAll(async () => {
		await server.ready();
		await database.connect();
	});

	beforeEach(async () => {
		const [testTask] = await Tasks.insertMany(task.getTasks(testLogin));
		taskId = testTask._id ?? '';
	});

	afterEach(async () => {
		await Tasks.deleteMany({});
	});

	afterAll(async () => {
		await server.close();
		await database.closeDatabase();
		await server.close();
	});

	it('Получение списка карточек', async () => {
		const token = auth.createToken({ login: testLogin }, server.config.SECRET_KEY);

		const { statusCode, payload } = await server.inject({
			method: 'GET',
			url: '/api/task/',
			payload: {},
			headers: {
				authorization: `Bearer ${token}`,
			},
		});

		expect(statusCode).toBe(200);
	});

	it('Добавление валидной карточки', async () => {
		const token = auth.createToken({ login: testLogin }, server.config.SECRET_KEY);

		const { statusCode, payload } = await server.inject({
			method: 'POST',
			url: '/api/task/',
			payload: {
				user: testLogin,
				title: 'Test insert title',
				description: 'Test insert description',
				whatWatch: 'serial',
				time: 4 * 60 * 1000,
				tags: ['insert', 'test'],
				completed: false,
			},
			headers: {
				authorization: `Bearer ${token}`,
			},
		});

		expect(statusCode).toBe(201);
	});

	it('Редактирование карточки', async () => {
		const token = auth.createToken({ login: testLogin }, server.config.SECRET_KEY);

		const { statusCode, payload } = await server.inject({
			method: 'PUT',
			url: `/api/task/${taskId}`,
			payload: {
				user: testLogin,
				title: 'Test update title',
				completed: false,
			},
			headers: {
				authorization: `Bearer ${token}`,
			},
		});

		expect(statusCode).toBe(204);
	});

	it('Удаление карточки карточки', async () => {
		const token = auth.createToken({ login: testLogin }, server.config.SECRET_KEY);

		const { statusCode, payload } = await server.inject({
			method: 'DELETE',
			url: `/api/task/${taskId}`,
			payload: {},
			headers: {
				authorization: `Bearer ${token}`,
			},
		});

		expect(statusCode).toBe(200);
		expect(JSON.parse(payload)).toStrictEqual({
			message: 'Task succesfull remove',
		});
	});
});
