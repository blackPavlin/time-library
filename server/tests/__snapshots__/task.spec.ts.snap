// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тест CRUD карточек Добавление валидной карточки 1`] = `
Object {
  "task": Object {
    "_id": "6155714acc14d50fb07d7a64",
    "completed": false,
    "description": "Test insert description",
    "tags": Array [
      "insert",
      "test",
    ],
    "time": 240000,
    "title": "Test insert title",
    "user": "admin@admin.com",
    "whatWatch": "serial",
  },
}
`;

exports[`Тест CRUD карточек Получение списка карточек 1`] = `
Object {
  "tasks": Array [
    Object {
      "_id": "615389870893021de40023ac",
      "completed": false,
      "description": "Test description film",
      "tags": Array [
        "test",
      ],
      "time": 60000,
      "title": "Test title silm",
      "user": "admin@admin.com",
      "whatWatch": "film",
    },
    Object {
      "_id": "615389870893021de40023ad",
      "completed": true,
      "description": "Test description serial",
      "tags": Array [
        "test",
      ],
      "time": 60000,
      "title": "Test title serial",
      "user": "admin@admin.com",
      "whatWatch": "serial",
    },
  ],
}
`;

exports[`Тест CRUD карточек Редактирование карточки 1`] = `
Object {
  "task": Object {
    "_id": "615389870893021de40023ac",
    "completed": false,
    "description": "Test description film",
    "tags": Array [
      "test",
    ],
    "time": 60000,
    "title": "Test update title",
    "user": "admin@admin.com",
    "whatWatch": "film",
  },
}
`;
